Add docs for c10::InferenceMode.

ailzhang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#57480 Add docs for c10::InferenceMode.**

Differential Revision: [D28164070](https://our.internmc.facebook.com/intern/diff/D28164070)

Commit:
Add docs for c10::InferenceMode.

[ghstack-poisoned]

ezyang commented on docs/cpp/source/notes/inference_mode.rst:
```diff
@@ -0,0 +1,122 @@
+with autograd (e.g. model training). Compared to ``NoGradMode``, code run
+under this mode gets better performance by disabling autograd related work like
+view tracking and version counter bumps. However, tensors created inside
+``c10::InferenceMode`` has more limitation interacting with autograd system as well.
```
when interacting

Commit:
Update on "Add docs for c10::InferenceMode."




[ghstack-poisoned]

albanD commented on docs/cpp/source/notes/inference_mode.rst:
```diff
@@ -0,0 +1,122 @@
+- are immutable outside ``InferenceMode``. So an error will be raised if you try to:
+  - mutate their data outside InferenceMode.
+  - mutate them into ``requires_grad=True`` outside InferenceMode.
+  To work around you can make a clone to get a normal tensor before mutating.
```
nit: make a clone outside of inferencemode

albanD commented on docs/cpp/source/notes/inference_mode.rst:
```diff
@@ -0,0 +1,122 @@
+
+Inside an ``InferenceMode`` block, we make the following performance guarantees:
+
+- Like ``NoGradMode``, all operations do not record ``grad_fn`` even if their ``requires_grad=True``.
```
nit missing "input"

albanD commented on docs/cpp/source/notes/inference_mode.rst:
```diff
@@ -0,0 +1,122 @@
+- Like ``NoGradMode``, all operations do not record ``grad_fn`` even if their ``requires_grad=True``.
+  This applies to both inference tensors and normal tensors.
+- View operations on inference tensors do not do view tracking. View and base inference tensors are
+  indistinguishable.
```
What do you mean by this?
Is it that view and non-views are indistinguishable?

albanD commented on docs/cpp/source/notes/inference_mode.rst:
```diff
@@ -0,0 +1,122 @@
+  indistinguishable.
+- Inplace operations on inference tensors are guaranteed not to do a version bump.
+
+For more implementation details of ``InferenceMode`` please see the `InferenceMode RFC <https://github.com/pytorch/rfcs/pull/17>`_.
```
Should the rfc be landed now that it is implemented? Or the goal is to keep them as PRs forever?

albanD commented on docs/cpp/source/notes/inference_mode.rst:
```diff
@@ -0,0 +1,122 @@
+which disables autograd, view tracking and version counter bumps. Unfortunately,
+current colloquial of this guard for inference workload is unsafe: it's possible to
+use ``AutoNonVariableTypeMode`` to bypass PyTorch's safety checks and result in
+silent wrong results, e.g. PyTorch throws an error when tensors saved for backwards
```
silently?

albanD commented on docs/cpp/source/notes/inference_mode.rst:
```diff
@@ -0,0 +1,122 @@
+When current users of ``AutoNonVariableTypeMode`` think about migrating, the following
+steps might help you decide the best alternatives:
+
+1. Users trying to run workload in inference only mode (like loading a pretrained JIT model and
```
Should we recommend loading the model itself in inference mode? So that the loaded weights are directly inference tensors?

albanD commented on docs/cpp/source/notes/inference_mode.rst:
```diff
@@ -0,0 +1,122 @@
+  - Both guards affects tensor execution process to skip work not related to inference, but ``InferenceMode``
+    also affects tensor creation while ``AutoNonVariableTypeMode`` doesn't. In other words, tensors created
+    inside ``InferenceMode`` are marked as inference tensors so that certain limitation can be applied when
+    they participate in autograd computation in the future.
```
nit: after exiting InferenceMode

albanD reviewed (approved):
LGTM!

Commit:
Update on "Add docs for c10::InferenceMode."


Differential Revision: [D28164070](https://our.internmc.facebook.com/intern/diff/D28164070)

[ghstack-poisoned]
